lib/irb/source_finder.rb

```diff
@@ -26,6 +31,13 @@ def find_source(signature)
       when /\A(?<owner>[A-Z]\w*(::[A-Z]\w*)*)#(?<method>[^ :.]+)\z/ # Class#method
         owner = eval(Regexp.last_match[:owner], context_binding)
         method = Regexp.last_match[:method]
+        if esses
+          super_owner = owner
+          s_count.times {|s| super_owner = super_owner.superclass}
+          if super_owner.respond_to?(:instance_method)
```

Comment:
`super_owner` is always `nil` or `Class` here.
The original intention of `if owner.respond_to?(:instance_method)` is to handle a case that owner is not a class.
```ruby
show_source RUBY_VERSION#method
```

How about moving `if esses` logic inside `if owner.respond_to?(:instance_method)` ?
```ruby
if owner.respond_to?(:instance_method)
  if esses
     ...
  end
  methods = owner.instance_methods + ...
end
```

Reply:
This is cool context you've given for the conditional.  Right now the `signature` getting passed to the method is a string including the `-s` . For example if I call `show_source Baz#foo -s` in irb.  This method will receive `"Baz#foo -s"` as the `signature` string and error out in the case statement.  That's why I have the parsing logic living at the top there.

Would you suggest that we change the `Regex` in the conditional for this method and do the parsing there? Then we could switch the order of the case statements and resolve it that way.

```ruby
case signature
      when /\A[A-Z]\w*(::[A-Z]\w*)*\z/ # Const::Name
        eval(signature, context_binding) # trigger autoload
        base = context_binding.receiver.yield_self { |r| r.is_a?(Module) ? r : Object }
        file, line = base.const_source_location(signature)
      when /\A(?<owner>[A-Z]\w*(::[A-Z]\w*)*)#(?<method>[^ :.]+)\z/ # Class#method
        owner = eval(Regexp.last_match[:owner], context_binding)
        method = Regexp.last_match[:method]
        if s_count
          super_owner = owner
          s_count.times {|s| super_owner = super_owner.superclass}
          if super_owner.respond_to?(:instance_method)
            owner = super_owner
          end
        end
        if owner.respond_to?(:instance_method)
          methods = owner.instance_methods + owner.private_instance_methods
          file, line = owner.instance_method(method).source_location if methods.include?(method.to_sym)
        end
```

Reply:
I was thinking of something like this. Parsing logic should be in top. Not change need for regexp.
```ruby
if owner.respond_to?(:instance_method) && s_count # to ensure owner is a class or module.
  # Maybe we need to check if owner is not a module, or support module too.
  super_owner = owner
  s_count.times do |s|
    # When super_owner reaches BasicObject, its superclass is nil.
    # To avoid calling nil.superclass, add `if super_owner.superclass` or `break unless super_owner` or use `&.`
    super_owner = super_owner.superclass
  end
  # super_owner is class or nil. To distinguish class and nil, respond_to?(:instance_method) is not needed.
  # Perhaps we can just `return unless super_owner` and show `Couldn't locate a definition`
  if super_owner needed.
    owner = super_owner
  end
end
if owner.respond_to?(:instance_method) # This check can be combined with the above one
  methods = owner.instance_methods + owner.private_instance_methods
  file, line = owner.instance_method(method).source_location if methods.include?(method.to_sym)
end
```
After two `if owner.respond_to?(:instance_method)` is combined, it will be like this, which I wanted to say in the previous comment
```ruby
if owner.respond_to?(:instance_method)
  if s_count # it was `if esses` before
     ...
  end
  methods = owner.instance_methods + ...
end
```
Or maybe using early return is more simple
```ruby
return unless owner.respond_to?(:instance_method)
if s_count
  ...
end
methods = ...
```

Reply:
Ah I understand now.  Sorry I misunderstood your original comment but I appreciate the second one for all of the cool ideas you mentioned!